Ripped from Fedora

* Mon Jun 14 2004 Alan Cox <alan@redhat.com>
- Re-arranged some totally bogus old bezerkly code that could
  segfault ftp on connection loss. (BZ #122295)

https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=122295

--- netkit-ftp-0.17/ftp/ftp.c.segv	2004-06-14 11:04:38.000000000 -0400
+++ netkit-ftp-0.17/ftp/ftp.c	2004-06-14 11:06:46.000000000 -0400
@@ -1790,6 +1799,10 @@
 			if (ptabflg)
 				code = -1;
 			lostpeer(0);
+			if (cout != NULL) {
+				fclose(cout);
+				cout = NULL;
+			}
 		}
 		(void) getreply(0);
 		(void) getreply(0);
@@ -1815,6 +1829,10 @@
 			INTOFF;
 			lostpeer(0);
 			INTON;
+			if (cout != NULL) {
+				fclose(cout);
+				cout = NULL;
+			}
 		}
 		else if (nfnd) {
 			(void) getreply(0);
@@ -1897,6 +1916,10 @@
 		if (ptabflg)
 			code = -1;
 		lostpeer(0);
+		if (cout != NULL) {
+			fclose(cout);
+			cout = NULL;
+		}
 	}
 	if (din && FD_ISSET(fileno(din), &mask)) {
 		while (read(fileno(din), buf, BUFSIZ) > 0)
--- netkit-ftp-0.17/ftp/main.c.segv	2004-06-14 11:03:18.000000000 -0400
+++ netkit-ftp-0.17/ftp/main.c	2004-06-14 11:03:42.000000000 -0400
@@ -235,8 +235,6 @@
 	if (connected) {
 		if (cout != NULL) {
 			shutdown(fileno(cout), 1+1);
-			fclose(cout);
-			cout = NULL;
 		}
 		if (data >= 0) {
 			shutdown(data, 1+1);
@@ -249,8 +247,6 @@
 	if (connected) {
 		if (cout != NULL) {
 			shutdown(fileno(cout), 1+1);
-			fclose(cout);
-			cout = NULL;
 		}
 		connected = 0;
 	}
